### PR TITLE
Fix(health): terraform conflict

### LINKED ISSUE #3355: Application 进行更新后 terrafrom component 无法通过 data-passing 拿到正确数据
**Describe the bug**
当我尝试对一个执行完成的 Application 进行资源调整是，发现 dependsOn 的组件没有获取到新的数据

**To Reproduce**

我的 comp
```
apiVersion: core.oam.dev/v1alpha2
kind: ComponentDefinition
metadata:
  name: demo-hello
  namespace: vela-system
  annotations:
    definition.oam.dev/description: Terraform configuration for demo Resource
  labels:
    type: terraform
spec:
  workload:
    definition:
      apiVersion: terraform.core.oam.dev/v1beta1
      kind: Configuration
  schematic:
    terraform:
      deleteResource: false
      configuration: |
        terraform {}

        variable "name" {
          type        = string
          description = "your name"
        }

        output "message" {
          value = "hello, ${var.name}"
        }
```

Application

```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: data-passing-test
spec:
  components:
    - name: data-passing-test-comp
      type: demo-hello
      outputs:
        - name: message
          valueFrom: output.status.apply.outputs.message.value
      properties:
        name: "vela!--"

    - name: data-passing-test-configmap
      type: raw
      dependsOn:
        - data-passing-test-comp
      inputs:
        - from: message
          parameterKey: properties.data.message
      properties:
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: data-passing-test-configmap
        type: kubernetes.io/opaque
        data: {}
```

首先使用 name: "vela" 执行，得到 configmap 如下：

```
apiVersion: v1
data:
  message: hello, vela
kind: ConfigMap
```

但是当我尝试更新 name: "vela!--" 时，configmap 并没有更新

**Expected behavior**

我期望 configmap 中的数据可以进行联动更新

**KubeVela Version**

version: v1.2.4

### PR DESCRIPTION
**Pull request by evanli18**
Signed-off-by: Evan Li <evan.li97@outlook.com>

### Description of your changes

"Fixes #3355":  Fix health state of the terrafrom comp

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/application/apply.go [hunk 1]
```diff
@@ -233,7 +233,9 @@ func (h *AppHandler) collectHealthStatus(ctx context.Context, wl *appfile.Worklo
 		if err := h.r.Client.Get(ctx, client.ObjectKey{Name: wl.Name, Namespace: namespace}, &configuration); err != nil {
 			return nil, false, errors.WithMessagef(err, "app=%s, comp=%s, check health error", appName, wl.Name)
 		}
-		if configuration.Status.Apply.State != terraformtypes.Available {
+		if appRev.Name != configuration.GetLabels()["app.oam.dev/appRevision"] ||
+			configuration.Status.Apply.State != terraformtypes.Available ||
+			configuration.Status.ObservedGeneration != configuration.Generation {
```
Comment (zzxwill) [inline pkg/controller/core.oam.dev/v1alpha2/application/apply.go:238]: Add an `&&` logic to check whether `configuration.Status.ObservedGeneration` is set.

Comment (evanli18) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:238]: That good. here we also need to compare apprev

Comment (zzxwill) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:238]: What will happen if we  don't compare apprev?

Comment (evanli18) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:238]: In kubebuilder, client.Get always gets data from the cache, so client.Get gets data from the cache immediately after applying object.

Ref: https://github.com/kubernetes-sigs/controller-runtime/issues/343

**Pull request closed without merging**